README.md
![](https://github.com/jcutler/Missing-E/raw/master/resources/logo/missinge48.png)
##Missing e: {browser extension/add-on for Tumblr!}

**Missing e** is a set of tools, features and interface changes for [Tumblr](http://www.tumblr.com). What started out as a bunch of userscripts has been incorporated into a full-fledged browser extension for _Chrome_, _Firefox_ and _Safari_!

**[GET IT](http://missing-e.com) - [FEATURES](http://missing-e.com/features) - [FAQ](http://missing-e.com/faq) - [TUMBLR](http://blog.missing-e.com) - [TWITTER](http://twitter.com/theMissinge)**

***

Jeremy Cutler ([cutlerish.tumblr.com](cutlerish.tumblr.com))  
Released under the [GPL Version 3 (or later) License](http://www.gnu.org/licenses/gpl.html)

[**Missing e** Change History](http://github.missing-e.com/changes)

_By downloading and using this extension, you agree to the terms and conditions set forth in the [**Missing e End User License Agreement**](http://missing-e.com/eula)_

If you enjoy this extension, please consider [donating](https://www.paypal.com/ca/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=EGQCRBB2BH5U8) to support future development.

***

###Note for Developers

The Firefox version of **Missing e** is built using the command-line [Mozilla Add-On SDK](https://addons.mozilla.org/en-US/developers/builder), which can be obtained [here](https://addons.mozilla.org/en-US/developers/builder).

The **Missing e** code has been reorganized as of version 2.3.0 so that most of the extension is platform-agnostic.

As such, I have included a [`setenvironment.sh`](https://github.com/jcutler/Missing-E/blob/master/setenvironment.sh) script that sets up Windows junctions and hard links so that shared folders and files can be added to all platform builds. This script was written for MINGW32 on a system with the Microsoft [`junction`](http://technet.microsoft.com/en-us/sysinternals/bb896768) utility (which you can [get here](http://technet.microsoft.com/en-us/sysinternals/bb896768)). On a *nix/OSX filesystem, you can simply replace all the `junction` and `fsutil hardlink` commands with `ln -s` to create symbolic links.


changes.md
---
title: Change History - Missing e - The browser extension for Tumblr!
---

<style type="text/css">
body {
font:13px Helvetica,Arial,sans-serif;
margin:20px;
padding:0 20px 10px;
max-width:900px;
border:1px solid #CCC;
border-radius:3px;
-webkit-box-shadow:rgba(0,0,0,0.07) 0 1px 2px inset;
-webkit-border-radius:3px;
-moz-border-radius:3px;
}

h1 {
font-size:170%;
border-top:4px solid #AAA;
padding-top:0.5em;
margin-top:1.5em;
}

h2, h3 {
margin:1em 0;
}

h2 {
font-size:130%;
margin-top:1.5em;
border-top:4px solid #E0E0E0;
padding-top:0.5em;
}

p {
margin:1em 0;
line-height:1.5em;
}

pre {
margin:1em 0;
font-size:12px;
background-color:#EEE;
border:1px solid #DDD;
padding:5px;
color:#444;
overflow:auto;
-webkit-box-shadow:rgba(0,0,0,0.07) 0 1px 2px inset;
-webkit-border-radius:3px;
-moz-border-radius:3px;
border-radius:3px;
}

code {
font-size:12px;
background-color:#f8f8ff;
color:#444;
padding:0 0.2em;
border:1px solid #DEDEDE;
}

pre code {
padding:0;
font-size:12px;
background-color:#EEE;
border:none;
}

a {
color:#4183c4;
}
</style>

#_Missing e_ Change History

    + indicates added features or functionality  
    - indicates features or functionality removed  
    ! indicates fixes to bugs and issues

##**2.14.3** (2012-12-04)
**_!_** Fix wonky icons for Bookmarker sidebar buttons

##**2.14.2** (2012-12-04)
**_!_** Fix wonky icons for added sidebar buttons

##**2.14.1** (2012-11-12)
**_!_** Add minimum Chrome version warning

##**2.14.0** (2012-11-12)
**_!_** Move to new Chrome extension format  
**_!_** Fix Ask Tweaks better answering functionality on Firefox  
**_!_** Minor functionality fixes  
**_!_** Fix avatar magnifier feature for follower/following pages  
**_!_** Prevent 'R' shortcut key from opening manual reblog page when using Quick Reblog  
**_+_** Add option to show tracked tags in sidebar

##**2.13.8** (2012-10-11)
**_!_** Correctly position Safe Dash sidebar button before sponsored radar posts

##**2.13.7** (2012-10-11)
**_!_** Fix tag reblogging from post permalink pages  
**_!_** Show blog selection in Quick Reblog drop-down on tag search result pages

##**2.13.6** (2012-10-11)
**_!_** Fix Goto Dash button on permalink pages

##**2.13.5** (2012-10-01)
**_!_** Fix issue with some Quick Reblog users experiencing failures

##**2.13.4** (2012-10-01)
**_!_** Fix issue with Quick Reblog menu not showing all blogs on secondary dash pages ending in "?lite"

##**2.13.3** (2012-09-28)
**_!_** Fix Safe Dash hiding for new Tumblr videos  
**_!_** Fix issue with multiple checkboxes showing up for inbox/post mass delete feature  
**_!_** Fix Reply Replies issue with some users unable to create reply posts for notes inside posts on the dashboard

##**2.13.2** (2012-09-26)
**_!_** Fix issue with Reblog Yourself adding empty/unusable blogs to posting drop-down with blogs using commas in blog title

##**2.13.1** (2012-09-07)
**_!_** Fix extra blog menu to show all blogs when reblogging from your own Tumblr  
**_!_** Clicking magnifier on avatars in dashboard notifications no longer opens new tab

##**2.13.0** (2012-09-07)
**_!_** Fix styling issues arising from new Tumblr dashboard layout  
**_!_** Fix placement of Mass Delete buttons  
**_!_** Fix Quick (1-click) Reblog menu not showing blog selector drop-down  
**_!_** Fix ability to reblog from one blog to that same blog

##**2.12.0** (2012-07-31)
**_!_** Fix for non-functioning Magnifier feature  
**_!_** Reduce white space when wrapping tags

##**2.11.0** (2012-07-14)
**_+_** Add X keyboard shortcut to enable/disable Safe Dash  
**_+_** Add option to display note count when hovering over posts in the Mass Editor  
**_!_** Fix incorrect note counts in dashboard notification previews for other languages  
**_!_** Additional fixes for post control icon alignment  
**_!_** Fix Ask Tweaks tagging of anonymous messages

##**2.10.6** (2012-07-11)
**_!_** Fix spacing of queue post arrow controls

##**2.10.5** (2012-07-11)
**_!_** Restore move to top of queue arrow display  
**_!_** Fix Smart Redirect issues for single-blog Tumblr accounts

##**2.10.4** (2012-07-10)
**_!_** Additional icon fixes

##**2.10.3** (2012-07-10)
**_!_** Fix post reblog button position for single-blog Tumblr accounts

##**2.10.2** (2012-07-10)
**_!_** Fix positioning of post control icons

##**2.10.1** (2012-07-10)
**_!_** Fix to allow tag-only posts  
**_!_** Disallow commas and semicolons in Bookmarker default bookmark format  
**_!_** Firefox: Fix reblog long text posts as text when reblogging from permalink page

##**2.10.0** (2012-06-29)
**_!_** Fix badly-formatted Quick Reblog drop-down menu

##**2.9.20** (2012-06-28)
**_!_** Fix problem with many users being unable to use Quick Reblog or Ask Tweaks advanced answering options

##**2.9.19** (2012-06-21)
**_!_** Fix issue with features not running on secondary following pages

##**2.9.18** (2012-06-21)
**_!_** Dutch localization corrections  
**_!_** Fix counts in added sidebar  
**_!_** Fix for Reply Replies in post notes

##**2.9.17** (2012-05-23)
**_+_** Dutch localization  
**_!_** Fix Reply Replies behaviour when replying to a post with no title beginning with a "Read More" break  
**_!_** Fix toggling of new "Send to Facebook" icon when creating new posts  
**_!_** Allow multiple quick reblogs of a single post without reloading  
**_!_** Fix note count display for dashboard notifications of video posts

##**2.9.16** (2012-05-18)
**_!_** Add slimming to dashboard blog menu for slim sidebar  
**_!_** Fix slim sidebar for new Find Blogs button  
**_!_** Fix popup box close button alignment

##**2.9.15** (2012-05-14)
**_!_** Fix clickability of magnifier button on Tumblr crush avatars  
**_!_** Fix Smart Redirect for single-blog Tumblr accounts

##**2.9.14** (2012-05-03)
**_!_** Correctly close extra settings tabs

##**2.9.13** (2012-05-01)
**_+_** Add Portuguese localization  
**_!_** Fix issue with automatic tagging of reblog posts not being added in some cases  

core/utils.js
/*
 * 'Missing e' Extension
 *
 * Copyright 2012, Jeremy Cutler
 * Released under the GPL version 3 licence.
 * SEE: license/GPL-LICENSE.txt
 *
 * This file is part of 'Missing e'.
 *
 * 'Missing e' is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * 'Missing e' is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with 'Missing e'. If not, see <http://www.gnu.org/licenses/>.
 */

(function(){

var utils = {
   defaultRetries: 4,
   minRetries: 0,
   maxRetries: 20,
   defaultFormat: "%Y-%m-%D %H:%i",
   
   debug: function(msg) {
      if (this.loggingLevel && this.loggingLevel >= 2) {
         console.debug(msg);
      }
   },
   
   log: function(msg) {
      if (this.loggingLevel && this.loggingLevel >= 1) {
         console.log(msg);
      }
   },

   urlPatterns: {
      "askForm":      /^ask_form\/[^\/]+(\/success)?[\/]?$/,
      "bookmarklet":  /^share(\/\w+)?[\/]?$/,
      "blog":         /^blog\/[^\/]+(\/search\/[^\/]+|\/processing)?(\/\d+)?[\/]?$/,
      "blogData":     /^blog\/[^\/]+\/((members|followers)(\/page\/\d+)?|settings)[\/]?$/,
      "crushes":      /^(blog\/[^\/]+\/)?new\/photo[\/]?$/,
      "dashboard":    /^(dashboard|show\/[^\/]*)([\/].*)?/,
      "dashboardOnly":/^dashboard([\/].*)?/,
      "drafts":       /^blog\/[^\/]+\/drafts[\/]?.*$/,
      "fanMail":      /^send(\/|_confirmation)?/,
      "followers":    /^blog\/[^\/]+\/followers[\/]?.*$/,
      "following":    /^following((\/page)?\/\d+|)?[\/]?$/,
      "iframe":       /^dashboard\/iframe/,
      "likes":        /^like(s(\/.*)?|d\/by\/[^\/]+(\/.*)?)$/,
      "massEditor":   /^mega-editor(\/[^\/]+(\/[0-9]+\/[0-9]+)?[\/]?|\/)?$/,
      "messages":     /^(blog\/[^\/]+\/)?(messages|submissions|inbox)[\/]?.*$/,
      "photoPost":    /^(blog\/[^\/]+\/)?new\/photo/,
      "post":         /^((blog\/[^\/]+\/)?new\/[^\/]+|edit\/\d+|share(\/\w+)?)[\/]?$/,
      "queue":        /^blog\/[^\/]+\/queue[\/]?.*$/,
      "reblog":       /^reblog\/\d+\/\w+(\/\w+)?[\/]?$/,
      "reply":        /^(blog\/[^\/]+\/)?new\/(text|photo)[\/]?$/,
      "settings":     /^blog\/[^\/]+\/settings[\/]?$/,
      "tagged":       /^tagged\/[^\/]+[\/]?.*$/,
      "textPost":     /^(blog\/[^\/]+\/)?new\/text/,
      "upload":       /^upload(\/image)?/
   },

   buildTimestamp: function(inStamp) {
      var i, x, stamp, tmp, tz, d, dq;
      var dt = {};
      var today = new Date();
      dt.day = today.getDay();
      stamp = inStamp.replace(/,/,'').split(" ");
      for (i=0; i<stamp.length; i++) {
         if (/^\d{4}$/.test(stamp[i])) {
            dt.year = parseInt(stamp[i], 10);
         }
         else if (/\d{1,2}:\d{2}$/.test(stamp[i])) {
            tmp = stamp[i].match(/(\d{1,2}):(\d{2})/);
            dt.hours = parseInt(tmp[1], 10);
            dt.minutes = parseInt(tmp[2].replace(/^0/,''), 10);
         }
         else if (/\d{1,2}:\d{2}(am|pm)/i.test(stamp[i])) {
            tmp = stamp[i].match(/(\d{1,2}):(\d{2})/);
            if (/pm$/i.test(stamp[i])) {
               dt.hours = parseInt(tmp[1], 10)+12;
               if (dt.hours === 24) { dt.hours = 12; }
            }
            else {
               dt.hours = parseInt(tmp[1], 10);
               if (dt.hours === 12) { dt.hours = 0; }
            }
            dt.minutes = parseInt(tmp[2].replace(/^0/,''), 10);
         }
         else if (/\d{1,2}\/\d{1,2}\/\d{4}/.test(stamp[i])) {
            tmp = stamp[i].match(/(\d{1,2})\/(\d{1,2})\/(\d{4})/);
            dt.date = parseInt(tmp[1], 10);
            dt.month = parseInt(tmp[2], 10);
            dt.year = parseInt(tmp[3], 10);
         }
         else if (/\d{1,2}(st|nd|rd|th)$/.test(stamp[i])) {
            tmp = stamp[i].match(/(\d{1,2})/);
            dt.date = parseInt(tmp[1], 10);
         }
         else if (/^[A-Za-z]+$/.test(stamp[i])) {
            var found = false;
            for (x=0; !found && x < this.getLocale("en")
                                       .monthsLong.length; x++) {
               if ((new RegExp(this.getLocale("en").monthsLong[x],"i"))
                     .test(stamp[i]) ||
                   (new RegExp(this.getLocale("en").monthsShort[x],"i"))
                     .test(stamp[i])) {
                  dt.month = x+1;
                  found = true;
               }
            }
            for (x=0; !found && x < this.getLocale("en")
                                       .daysLong.length; x++) {
               if ((new RegExp(this.getLocale("en").daysLong[x],"i"))
                     .test(stamp[i]) ||
                   (new RegExp(this.getLocale("en").daysShort[x], "i"))
                     .test(stamp[i])) {
                  dt.date = (x - dt.day - 7) % 7;
                  found = true;
               }
            }
         }
      }
      if (dt.year) {
         d = new Date(dt.month + "/" + dt.date + "/" + dt.year + " " +
                         dt.hours + ":" + dt.minutes + ":00");
         return Math.round(d.getTime()/1000);
      }
      if (!dt.year) {
         if (dt.month) {
            dq = new Date(dt.month + "/" + dt.date + "/" +
                              today.getFullYear() + " " + dt.hours +
                              ":" + dt.minutes + ":00");
            /* If more than a day ahead, month is in previous year */
            if (dq > today + 86400000) {
               dt.year = today.getFullYear() - 1;
            }
            else {
               dt.year = today.getFullYear();
            }
            d = new Date(dt.month + "/" + dt.date + "/" + dt.year +
                             " " + dt.hours + ":" + dt.minutes + ":00");
            return Math.round(d.getTime()/1000);
         }
         else if (dt.date < 0) {
            dq = new Date((today.getMonth()+1) + "/" +
                              today.getDate() + "/" +
                              today.getFullYear() + " " + dt.hours +
                              ":" + dt.minutes + ":00 UTC");
            /* If more than a month ahead, wrapped backwards across year */
            if (dq > today + 2764800000) {
               today.setFullYear(dt.year);
            }
            /* If more than a day ahead, wrapped backwards across month */
            if (dq > today + 86400000) {
               today.setMonth(today.getMonth()-1);
            }
            today = new Date(today.valueOf()+86400000*dt.date);
            dt.date = today.getDate();
            dt.month = today.getMonth()+1;
            dt.year = today.getFullYear();
            d = new Date(dt.month + "/" + dt.date + "/" + dt.year +
                             " " + dt.hours + ":" + dt.minutes + ":00");
            return Math.round(d.getTime()/1000);
         }
         else if (!dt.date) {
            dt.year = today.getFullYear();
            dt.month = today.getMonth()+1;
            dt.date = today.getDate();
            d = new Date(dt.month + "/" + dt.date + "/" + dt.year + " " +
                           dt.hours + ":" + dt.minutes + ":00");
            return Math.round(d.getTime()/1000);
         }
      }
      return null;
   },

   createTag: function(lang, txt) {
      var theTag = document.createElement("div");
      var theSpan = document.createElement("span");
      var theA = document.createElement("a");
      theTag.className = "token";
      theSpan.className = "tag";
      theSpan.textContent = txt;
      theTag.appendChild(theSpan);
      theA.title = MissingE.getLocale(lang).removeTag;
      theA.href = "#";
      theA.textContent = "x";
      theA.addEventListener("click", function(e){
         var atag = this.parentNode;
         var tagbox = atag.parentNode;
         tagbox.removeChild(atag);
         document.getElementById("tag_editor_input").blur();
         e.preventDefault();
      }, false);
      theTag.appendChild(theA);
      return theTag;
   },

   escapeHTML: function(str) {
      return str.replace(/&/g,'&amp;').replace(/"/g,'&quot;')
               .replace(/>/,'&gt;').replace(/</,'&lt;');
   },

   getBookmarkerFormat: function(d, user, format, lang) {
      var ret = this.getFormattedDate(d, format, lang);
      return ret.replace(/%u/g,user);
   },

   getFormattedDate: function(d, format, lang) {
      var ret = format;
      ret = ret.replace(/%Y/g,d.getFullYear())
               .replace(/%y/g,(d.getFullYear()%100))
               .replace(/%M/g,this.getLocale(lang).monthsShort[d.getMonth()])
               .replace(/%B/g,this.getLocale(lang).monthsLong[d.getMonth()])
               .replace(/%w/g,this.getLocale(lang).daysShort[d.getDay()])
               .replace(/%W/g,this.getLocale(lang).daysLong[d.getDay()])
               .replace(/%m/g,this.zeroPad(d.getMonth()+1,2))
               .replace(/%n/g,(d.getMonth()+1))
               .replace(/%D/g,this.zeroPad(d.getDate(),2))
               .replace(/%d/g,d.getDate())
               .replace(/%G/g,this.zeroPad((d.getHours()%12===0 ?
                                       "12" : d.getHours()%12),2))
               .replace(/%g/g,(d.getHours()%12===0 ? "12" : d.getHours()%12))
               .replace(/%H/g,this.zeroPad(d.getHours(),2))
               .replace(/%h/g,d.getHours())
               .replace(/%i/g,this.zeroPad(d.getMinutes(),2))
               .replace(/%s/g,this.zeroPad(d.getSeconds(),2))
               .replace(/%A/g,(d.getHours() < 12 ? "AM" : "PM"))
               .replace(/%a/g,(d.getHours() < 12 ? "am" : "pm"));
      return ret;
   },

   getLocale: function(lang) {
      if (typeof lang !== "string") {
         lang = "en";
      }
      lang = lang.toLowerCase();
      if (this.locale.hasOwnProperty(lang) &&
          this.locale[lang] !== false) {
         return this.locale[lang];
      }
      else {
         if (!this.locale.hasOwnProperty(lang)) {
             this.locale[lang] = false;
            MissingE.debug("Warning: Localization not found for language '" +
                        lang + "'");
         }
         return this.locale.en;
      }
   },

   getStorage: function(key, defVal) {
      var retval = localStorage[key];
      if (retval === undefined || retval === null || retval === "") {
         return defVal;
      }
      else {
         if (/[^\d]/.test(retval)) {
            return retval;
         }
         else {
            return parseInt(retval, 10);
         }
      }
   },

   isEDT: function(year, month, date, hours) {
      var add;
      var start = new Date("03/01/" + year + " 07:00:00 UTC");
      var day = start.getDay();
      if (day === 0) { add = 7; }
      else { add = 14-day; }
      start.setUTCDate(start.getUTCDate()+add);
      var end = new Date("11/01/" + year + " 06:00:00 UTC");
      day = end.getDay();
      if (day === 0) { add = 0; }
      else { add = 7-day; }
      end.setUTCDate(end.getUTCDate()+add);
      if (month>start.getUTCMonth()+1 && month<end.getUTCMonth()+1) {
         return true;
      }
      else if (month === start.getUTCMonth()+1) {
         if (date === start.getUTCDate()) {
            if (hours >= 2) {
               return true;
            }
            else {
               return false;
            }
         }
         else {
            return date > start.getUTCDate();
         }
      }
      else if (month === end.getUTCMonth()+1) {
         if (date === end.getUTCDate()) {
            if (hours >= 2) {
               return false;
            }
            else {
               return true;
            }
         }
         else {
            return date < end.getUTCDate();
         }
      }
      else {
         return false;
      }
   },

   isEDTfromUTC: function(dt) {
      var add;
      var start = new Date("03/01/" + dt.getUTCFullYear() + " 07:00:00 UTC");
      var day = start.getDay();
      if (day === 0) { add = 7; }
      else { add = 14-day; }
      start.setUTCDate(start.getUTCDate()+add);
      var end = new Date("11/01/" + dt.getUTCFullYear() + " 06:00:00 UTC");
      day = end.getDay();
      if (day === 0) { add = 0; }
      else { add = 7-day; }
      end.setUTCDate(end.getUTCDate()+add);
      return (dt>=start && dt<end);
   },

   isSameDay: function(dt) {
      if (typeof dt === "string") {
         dt = parseInt(dt,10);
      }
      var now = new Date();
      var then = new Date(dt);
      if (now > then + 86400000) {
         return false;
      }
      return now.getFullYear() === then.getFullYear() &&
             now.getMonth() === then.getMonth() &&
             now.getDate() === then.getDate();
   },

   isTumblrURL: function(fullURL, matches) {
      if (!/^http:\/\/www\.tumblr\.com\//.test(fullURL)) {
         return false;
      }
      else if (!matches) {
         return true;
      }
      var i;
      var url = fullURL.replace(/^http:\/\/www\.tumblr\.com\//,'')
                       .replace(/\?.*$/,'').replace(/#.*$/,'');
      for (i=0; i<matches.length; i++) {
         if (!this.urlPatterns.hasOwnProperty(matches[i])) {
            debug("Invalid Tumblr URL pattern: '" + matches[i] + "'");
         }
         else {
            if (this.urlPatterns[matches[i]].test(url)) {
               return true;
            }
         }
      }
      return false;
   },

   randomRange: function(from, to) {
      return Math.floor(Math.random() * (to - from + 1) + from);
   },

   setStorage: function(key, val) {
      localStorage[key] = val;
   },

   unescapeHTML: function(str) {
      var ret = str;
      while (/&#[0-9]*;/.test(ret)) {
         var entity = ret.match(/&#([0-9]+);/);
         var symbol = String.fromCharCode(entity[1]);
         ret = ret.replace(entity[0], symbol);
      }
      return ret.replace(/&amp;/g,'&').replace(/&quot;/g,'"')
               .replace(/&gt;/g,'>').replace(/&lt;/,'<');
   },

   versionCompare: function(v1, v2) {
      if (!v1 && !v2) { return 0; }
      else if (!v1) { return -1; }
      else if (!v2) { return 1; }
      else {
         var i;
         var ver1 = v1.split('.');
         var ver2 = v2.split('.');
         var len = ver1.length >= ver2.length ? ver1.length : ver2.length;
         for (i=0; i<len; i++) {
            if (i >= ver1.length && ver2[i] !== '0') { return -1; }
            else if (i >= ver2.length && ver1[i] !== '0') { return 1; }
            else {
               ver1[i] = parseInt(ver1[i]);
               ver2[i] = parseInt(ver2[i]);
            }
            if (ver1[i] > ver2[i]) { return 1; }
            if (ver2[i] > ver1[i]) { return -1; }
         }
         return 0;
      }
   },

   zeroPad: function(num, len) {
      var ret = "";
      ret += num;
      while (ret.length < len) { ret = "0" + ret; }
      return ret;
   }
};

if (typeof window === "undefined" &&
    typeof require === "function" &&
    require("api-utils/xul-app")) {
   // No logging or debugging output in Firefox released add-on
   if (utils.loggingLevel === undefined) {
      utils.loggingLevel = 0;
   }
   exports.utils = utils;
}
else if (!MissingE.hasOwnProperty("_utilsLoaded") ||
         !MissingE._utilsLoaded) {
   var i;
   if (utils.loggingLevel === undefined) {
      if (extension.isFirefox) {
         // no logging or debugging output in Firefox released add-on
         utils.loggingLevel = 0;
      }
      else {
         // other browsers will display logging but not debugging output
         utils.loggingLevel = 1;
      }
   }
   MissingE._utilsLoaded = true;
   for (i in utils) {
      if (utils.hasOwnProperty(i)) {
         MissingE[i] = utils[i];
      }
   }

   Array.prototype.shuffle = function () {
      var s = [];
      while (this.length) {
         s.push(this.splice(Math.random() * this.length, 1)[0]);
      }
      while (s.length) {
         this.push(s.pop());
      }
      return this;
   };
}

}());
